src/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Razor/RazorHelperUpdater.cs

```diff
@@ -0,0 +1,153 @@
+            _logger.LogInformation("Found @helper functions in {Count} documents", inputsWithHelpers.Count);
+            return new FileUpdaterResult(inputsWithHelpers.Any(), inputsWithHelpers);
+        }
+
+        /// <summary>
+        /// Upgrades a set of RazorCodeDocuments to replace @helper functions with local methods.
+        /// </summary>
+        /// <param name="context">The upgrade context being upgraded.</param>
+        /// <param name="inputs">The Razor code documents to upgrade.</param>
+        /// <param name="token">A cancellation token.</param>
+        /// <returns>A FileUpdaterResult indicating whether the opertaion succeeded and which documents (if any) were updated.</returns>
+        public async Task<IUpdaterResult> ApplyAsync(IUpgradeContext context, ImmutableArray<RazorCodeDocument> inputs, CancellationToken token)
+        {
+            if (context is null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            // Track which documents are updated for inclusion in the updater result returned later
+            var updatedDocuments = new List<string>();
+            foreach (var document in inputs)
+            {
+                var path = document.Source.FilePath;
+                var helperReplacements = await _helperMatcher.GetHelperReplacementsAsync(document).ConfigureAwait(false);
+
+                if (helperReplacements.Any())
+                {
+                    // If the document contains a @helper, construct the new document text by removing the replacement's old text
+                    // and inserting the new text in its place.
+                    updatedDocuments.Add(path);
+                    var newText = new StringBuilder(File.ReadAllText(path));
+                    _logger.LogInformation("Replacing {HelperCount} @helper functions in {FilePath}", helperReplacements.Count(), path);
+                    foreach (var helper in helperReplacements.OrderByDescending(h => h.Offset))
+                    {
+                        helper.Apply(newText);
+                    }
+
+                    // Check whether the document has a using statement for HelperResult's namespace (Microsoft.AspNetCore.Mvc.Razor) and
+                    // add one if none is present. Again, this is done with string manipulation since there's no good API for updating
+                    // Razor syntax.
+                    if (!HasMvcRazorUsing(document))
+                    {
+                        _logger.LogInformation("Adding an import statement for the '{namespace}' namespace to {FilePath}", HelperResultNamespace, path);
+                        newText.Insert(0, $"@using {HelperResultNamespace}{Environment.NewLine}");
+                    }
+
+                    _logger.LogDebug("Writing updates to {FilePath} to replace @helper functions", path);
+                    File.WriteAllText(path, newText.ToString());
```

Comment:
Can't we use a SourceText or something so we go through that abstraction rather than directly to the File IO?

Reply:
That would be better. We'd still need a TextWriter to give the SourceText, though. We could maybe include a file IO abstraction in DI. I'm going to leave that for a future update, though, to keep this one focused.